src/main/java/com/minecolonies/coremod/client/gui/townhall/WindowMainPage.java

```diff
@@ -64,111 +101,376 @@ public WindowMainPage(final BuildingTownHall.View building)
+        if (key.equals(COM_MINECOLONIES_COREMOD_GUI_HIRING_OFF))
+        {
+            button.setText(Component.translatable(COM_MINECOLONIES_COREMOD_GUI_HIRING_ON));
+            toggle = true;
+        }
+        else
+        {
+            button.setText(Component.translatable(COM_MINECOLONIES_COREMOD_GUI_HIRING_OFF));
+            toggle = false;
+        }
+        Network.getNetwork().sendToServer(new ToggleHousingMessage(this.building.getColony(), toggle));
     }
 
     /**
-     * Executed when <code>WindowTownHall</code> is opened. Does tasks like setting buttons.
+     * Toggles citizens moving in. Off means citizens stop moving in.
+     *
+     * @param button the pressed button.
      */
-    @Override
-    public void onOpened()
+    private void toggleMoveIn(@NotNull final Button button)
     {
-        super.onOpened();
-        fillAlliesAndFeudsList();
+        String key = button.getText().getContents() instanceof TranslatableContents ? ((TranslatableContents) button.getText().getContents()).getKey() : button.getTextAsString();
 
-        title.setText(Component.literal(building.getColony().getName()));
+        final boolean toggle;
+        if (key.equals(OFF_STRING))
+        {
+            button.setText(Component.translatable(ON_STRING));
+            toggle = true;
+        }
+        else
+        {
+            button.setText(Component.translatable(OFF_STRING));
+            toggle = false;
+        }
+        Network.getNetwork().sendToServer(new ToggleMoveInMessage(this.building.getColony(), toggle));
+    }
 
-        if (building.getColony().getMercenaryUseTime() != 0
-              && building.getColony().getWorld().getGameTime() - building.getColony().getMercenaryUseTime() < TICKS_FOURTY_MIN)
+    /**
+     * Toggles printing progress.
+     *
+     * @param button the button to toggle.
+     */
+    private void togglePrintProgress(@NotNull final Button button)
+    {
+        String key = button.getText().getContents() instanceof TranslatableContents ? ((TranslatableContents) button.getText().getContents()).getKey() : button.getTextAsString();
+
+        if (key.equals(OFF_STRING))
         {
-            findPaneOfTypeByID(BUTTON_MERCENARY, Button.class).disable();
+            button.setText(Component.translatable(ON_STRING));
+        }
+        else
+        {
+            button.setText(Component.translatable(OFF_STRING));
         }
+        Network.getNetwork().sendToServer(new ToggleHelpMessage(this.building.getColony()));
+    }
+
+    /**
+     * Switch the structure style pack.
+     */
+    private void switchPack()
+    {
+        new WindowSwitchPack(() -> {
+            building.getColony().setStructurePack(StructurePacks.selectedPack.getName());
+            Network.getNetwork().sendToServer(new ColonyStructureStyleMessage(building.getColony(), StructurePacks.selectedPack.getName()));
+            return new WindowMainPage((BuildingTownHall.View) this.building);
+        }).open();
```

Comment:
You can try to use the layering system `openAsLayer`, which avoids open() machinery